Lib/Stack.hpp

```diff
@@ -613,6 +613,21 @@ class Stack
     const Stack& _stack;
   };
 
+  class ConstIterator : public ConstRefIterator {
+  public:
+    ConstIterator(Stack const& s) : ConstRefIterator(s) {}
+    DECL_ELEMENT_TYPE(C);
+    C next() { return ConstRefIterator::next(); }
+  };
+
+  RefIterator iter() &
+  { return RefIterator(*this); }
+
+  ConstRefIterator iter() const&
+  { return ConstRefIterator(*this); }
+
+
```

Comment:
Could you explain what's going on here for me? Not sure how it relates to current change.

Reply:
Soooo the issue was that the `Stack` iterators were not consistent with vampire's (informal) iterator interface. 
That is `Stack::Iterator` used to have `DECL_ELEMENT_TYPE(C)` but the function `next` did return `C&`. 
This means in cases where `C` was not copy constructible, one couldn't use our iterator adaptors (`MappingIterator` etc). Still some parts of the codebase (i.e. the implemenation of constraints in `Options.hpp`) relied on the iterators returning a reference on next (because they dealt with `unique_ptr` which are not copyable). In order to resolve this I created two versions of iterators: `Stack::Iterator` and `Stack::RefIterator` (where one is just a copying wrapper around the other one), that both conform to the iterator interface.
(And of course the same for `ConstRefIterator` vs `ConstIterator`).

(Sorry that should have maybe been a separate PR? Just changed it bc I needed it.)

Reply:
Oh, right. Yeah, that's gross - good catch, no objections here.